File: apps/api/src/sandbox/managers/actions/sandbox-destroy.action.ts
import { Injectable } from "@nestjs/common";
import { InjectRepository } from "@nestjs/typeorm";
import { Repository } from "typeorm";
import { ExecutorAdapterFactory } from "../../adapter/adapter";
import { Sandbox } from "../../entities/sandbox.entity";
import { ExecutorState } from "../../enums/executor-state.enum";
import { SandboxState } from "../../enums/sandbox-state.enum";
import { ExecutorService } from "../../services/executor.service";
import { ToolboxService } from "../../services/toolbox.service";
import { DONT_SYNC_AGAIN, SandboxAction, SYNC_AGAIN, SyncState } from "./sandbox.action";

@Injectable()
export class SandboxDestroyAction extends SandboxAction {
  constructor(
    protected executorService: ExecutorService,
    protected executorAdapterFactory: ExecutorAdapterFactory,
    @InjectRepository(Sandbox)
    protected sandboxRepository: Repository<Sandbox>,
    protected toolboxService: ToolboxService
  ) {
    super(executorService, executorAdapterFactory, sandboxRepository, toolboxService);
  }

  async run(sandbox: Sandbox): Promise<SyncState> {
    if (sandbox.state === SandboxState.ARCHIVED) {
      await this.updateSandboxState(sandbox.id, SandboxState.DESTROYED);
      return DONT_SYNC_AGAIN;
    }

    const executor = await this.executorService.findOne(sandbox.executorId);
    if (executor.state !== ExecutorState.READY) {
      return DONT_SYNC_AGAIN;
    }

    const executorAdapter = await this.executorAdapterFactory.create(executor);

    switch (sandbox.state) {
      case SandboxState.DESTROYED:
        return DONT_SYNC_AGAIN;
      case SandboxState.DESTROYING: {
        // check if sandbox is destroyed
        try {
          const sandboxInfo = await executorAdapter.sandboxInfo(sandbox.id);
          if (
            sandboxInfo.state === SandboxState.DESTROYED ||
            sandboxInfo.state === SandboxState.ERROR
          ) {
            await executorAdapter.removeDestroyedSandbox(sandbox.id);
          }
        } catch (e) {
          //  if the sandbox is not found on executor, it is already destroyed
          if (!e.response || e.response.status !== 404) {
            throw e;
          }
        }

        await this.updateSandboxState(sandbox.id, SandboxState.DESTROYED);
        return SYNC_AGAIN;
      }
      default: {
        // destroy sandbox
        try {
          const sandboxInfo = await executorAdapter.sandboxInfo(sandbox.id);
          if (sandboxInfo?.state === SandboxState.DESTROYED) {
            await this.updateSandboxState(sandbox.id, SandboxState.DESTROYING);
            return SYNC_AGAIN;
          }
          await executorAdapter.destroySandbox(sandbox.id);
        } catch (e) {
          //  if the sandbox is not found on executor, it is already destroyed
          if (e.response.status !== 404) {
            throw e;
          }
        }
        await this.updateSandboxState(sandbox.id, SandboxState.DESTROYING);
        return SYNC_AGAIN;
      }
    }
  }
}

File: crates/api/src/services/sandbox_destroy_action.rs
// Copyright (c) 2026 Snapflow. All rights reserved.
//
// Snapflow is licensed under the GNU Affero General Public License v3.0.
// You may obtain a copy of the license at https://www.gnu.org/licenses/agpl-3.0.html
//
// SPDX-License-Identifier: AGPL-3.0

use crate::executor;
use crate::infra::Infra;
use crate::models::Sandbox;
use crate::repositories;
use snapflow_errors::Result;
use snapflow_models::{BackupState, ExecutorState, SandboxState};

use super::sandbox_actions::{StateUpdate, SyncState, update_sandbox_state};

pub async fn run(infra: &Infra, sandbox: &Sandbox) -> Result<SyncState> {
    if sandbox.state == SandboxState::Archived {
        update_sandbox_state(
            infra,
            sandbox.id,
            SandboxState::Destroyed,
            StateUpdate {
                backup_state: Some(BackupState::None),
                ..Default::default()
            },
        )
        .await?;
        return Ok(SyncState::Done);
    }

    let Some(executor_id) = sandbox.executor_id else {
        update_sandbox_state(
            infra,
            sandbox.id,
            SandboxState::Destroyed,
            StateUpdate {
                backup_state: Some(BackupState::None),
                ..Default::default()
            },
        )
        .await?;
        return Ok(SyncState::Done);
    };

    let Some(exec) = repositories::executor::find_by_id(&infra.pool, executor_id).await? else {
        update_sandbox_state(
            infra,
            sandbox.id,
            SandboxState::Destroyed,
            StateUpdate {
                backup_state: Some(BackupState::None),
                ..Default::default()
            },
        )
        .await?;
        return Ok(SyncState::Done);
    };

    if exec.state != ExecutorState::Ready {
        return Ok(SyncState::Again);
    }

    let adapter = executor::create_adapter(&exec)?;
    let sandbox_id_str = sandbox.id.to_string();

    match sandbox.state {
        SandboxState::Destroyed => Ok(SyncState::Done),
        SandboxState::Destroying => {
            match adapter.sandbox_info(&sandbox_id_str).await {
                Ok(info) => {
                    if matches!(info.state, SandboxState::Destroyed | SandboxState::Error)
                        && let Err(e) = adapter.remove_destroyed_sandbox(&sandbox_id_str).await
                    {
                        tracing::warn!(sandbox_id = %sandbox_id_str, error = %e, "failed to remove destroyed sandbox from executor");
                    }
                }
                Err(e) => {
                    let msg = e.to_string();
                    if !msg.contains("404") && !msg.contains("not found") {
                        return Err(e);
                    }
                }
            }

            update_sandbox_state(
                infra,
                sandbox.id,
                SandboxState::Destroyed,
                StateUpdate::default(),
            )
            .await?;
            Ok(SyncState::Done)
        }
        _ => {
            let already_destroyed = match adapter.sandbox_info(&sandbox_id_str).await {
                Ok(info) => {
                    if info.state == SandboxState::Destroyed {
                        true
                    } else {
                        adapter.destroy_sandbox(&sandbox_id_str).await?;
                        false
                    }
                }
                Err(e) => {
                    let msg = e.to_string();
                    if msg.contains("404") || msg.contains("not found") {
                        true
                    } else {
                        return Err(e);
                    }
                }
            };

            if already_destroyed {
                if let Err(e) = adapter.remove_destroyed_sandbox(&sandbox_id_str).await {
                    tracing::warn!(sandbox_id = %sandbox_id_str, error = %e, "failed to remove already-destroyed sandbox from executor");
                }
                update_sandbox_state(
                    infra,
                    sandbox.id,
                    SandboxState::Destroyed,
                    StateUpdate::default(),
                )
                .await?;
                Ok(SyncState::Done)
            } else {
                update_sandbox_state(
                    infra,
                    sandbox.id,
                    SandboxState::Destroying,
                    StateUpdate::default(),
                )
                .await?;
                Ok(SyncState::Again)
            }
        }
    }
}
